CT: Update auto-complete search functionality to use REDIS #4096

zurbergram opened the pull request:
## Description of change
As a developer, I need to update the auto-complete search functionality of the comparison tool to use REDIS, so that it uses cached data for search functionality to prevent performance issues when completing searches.

These changes add REDIS caching to all calls to gibct-data-service

https://github.com/department-of-veterans-affairs/va.gov-team/issues/4046

## Testing done
- [x] /v0/gi/institutions/autocomplete?term=college
  - Use autocomplete for non-VetTec
- [x] /v0/gi/institutions/search
  - Search for a non-VetTec institution
- [x] /v0/gi/institutions/show
  - Load an institution's profile
- [x] /v0/gi/institutions/children
  - Not used
- [x] /v0/gi/institution_programs/autocomplete?term=college
  - Use autocomplete for VetTec
- [x] /v0/gi/institution_programs/search
  - Search for a VetTec program
- [x] /v0/gi/calculator_constants/
  - Load search results
- [x] /v0/gi/zipcode_rates/
  - Load an institution's profile

## Testing planned
Smoke testing of https://staging.va.gov/gi-bill-comparison-tool/ 

## Acceptance Criteria (Definition of Done)
- [x] 1. The non VETTEC institution auto-complete uses “cache-aside” to perform searches.
- [x] 2. The VETTEC provider auto-complete uses “cache-aside” to perform searches.

#### Unique to this PR
<!-- This would be a good place to include feature flag check item and info, specific dashboards and instrumentation check item and info -->
- [ ] _Replace this line with individual tasks unique to your PR_

#### Applies to all PRs

- [x] Appropriate logging
- [x] Swagger docs have been updated, if applicable
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [x] Provide which alerts would indicate a problem with this functionality (if applicable)

jholton commented:
Caching these responses is a great idea.

Can you refactor the implementation to more closely follow the pattern that the other Redis-backed services use?  See `Vet360Redis` and subclasses, `EMISRedis` and subclasses, and `OktaRedis` and subclasses.
I think `Vet360Redis::ReferenceData` is a good example to start with.

Perhaps a `GIDSRedis` class with `Institution`, `InstitutionProgram`, `CalculatorConstant` and `ZipCodeRate` subclasses.
Then in the controller you can just call methods like 
```ruby
client = GIDSRedis::Institution.new
client.autocomplete(terms)
```

That'll be much easier to understand at first read, and follows established patterns.


In these models I would like to see use of `GIDS` instead of `GI` in naming.
I know you're following established naming, but I think `GI` is too generic.  I would have advocated for the more expressive `GIDS` in the `lib` files/directories if I had been working on the project back when the code was originally added.  
Since adding `GIDS` integration, we have added a `GI Bill Feedback Tool` form.  That and any other features we might add in the future that contains 'GI' in the title only increase the `GI` naming confusion.
(If you're adding classes/modules in the existing `lib/gi/` directory I understand that you have to use `GI` to adhere to naming conventions.)

annaswims commented on lib/gi/configuration.rb:
```diff
@@ -8,8 +8,6 @@
 
 module GI
   class Configuration < Common::Client::Configuration::REST
-    self.read_timeout = Settings.gids.read_timeout || 1
```
We cannot allow you to increase the timeout back to 15s  for production. If you feel the need to override this for staging you may  do it [here](https://github.com/department-of-veterans-affairs/devops/blob/master/ansible/deployment/config/vets-api/staging-settings.local.yml.j2#L72)

Why:
https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/pull/17/files?short_path=7e8cf21#diff-7e8cf218cb2da0710948af2719b5e320

zurbergram commented:
> Caching these responses is a great idea.
> 
> Can you refactor the implementation to more closely follow the pattern that the other Redis-backed services use? See `Vet360Redis` and subclasses, `EMISRedis` and subclasses, and `OktaRedis` and subclasses.
> I think `Vet360Redis::ReferenceData` is a good example to start with.
> 
> Perhaps a `GIDSRedis` class with `Institution`, `InstitutionProgram`, `CalculatorConstant` and `ZipCodeRate` subclasses.
> Then in the controller you can just call methods like
> 
> ```ruby
> client = GIDSRedis::Institution.new
> client.autocomplete(terms)
> ```
> 
> That'll be much easier to understand at first read, and follows established patterns.
> 
> In these models I would like to see use of `GIDS` instead of `GI` in naming.
> I know you're following established naming, but I think `GI` is too generic. I would have advocated for the more expressive `GIDS` in the `lib` files/directories if I had been working on the project back when the code was originally added.
> Since adding `GIDS` integration, we have added a `GI Bill Feedback Tool` form. That and any other features we might add in the future that contains 'GI' in the title only increase the `GI` naming confusion.
> (If you're adding classes/modules in the existing `lib/gi/` directory I understand that you have to use `GI` to adhere to naming conventions.)

Made changes
- renamed as much as was possible outside from "gi" to be "gids"
- controllers can now call `service.method_name(scrubbed_params)` where method_name is a method in `lib/gi/client.rb`

jholton commented on app/controllers/gids_controller.rb:
```diff
@@ -1,14 +1,14 @@
+require 'gids_redis/Gids'
 
-class GIController < ApplicationController
+class GidsController < ApplicationController
```
Can we upcase `Gids` -> `GIDS` wherever possible (like `GI` is upcase in the 'before' of this diff).  You'll need to add `GIDS` to `config/initializers/inflections.rb`

zurbergram replied:
changed

jholton commented on app/controllers/gids_controller.rb:
```diff
@@ -1,14 +1,14 @@
 # frozen_string_literal: true
 
-require 'gi/client'
+require 'gids_redis/Gids'
```
Is this necessary?

zurbergram replied:
Locally I get below, even after `make rebuild; make up` when removing that line

```"exception": "uninitialized constant GidsController::Gids",
"backtrace": [
"/srv/vets-api/src/app/controllers/gids_controller.rb:9:in `service'",
"/srv/vets-api/src/app/controllers/v0/gids/calculator_constants_controller.rb:7:in `index'",```

zurbergram replied:
was able to remove

jholton commented on app/models/gids_redis/gids.rb:
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require 'common/models/redis_store'
+require 'common/models/concerns/cache_aside'
```
I don't think these are necessary once you fix the name/path of this class.

zurbergram replied:
removed

zurbergram commented on app/models/gids_redis/gids.rb:
```diff
@@ -0,0 +1,55 @@
+require 'common/models/concerns/cache_aside'
+
+# Facade for GI.
+class Gids < Common::RedisStore
```
fixed

jholton commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,55 @@
+
+  # @return the response returned from GI Client
+  def gi_response
+    @gi_response ||= response_from_redis_or_service
```
You definitely don't want to memoize this.
```
service = GIDSRedis.new
service.get_calculator_constants
# => returns calculator constants data
service.get_institution_program_autocomplete_suggestions
# => returns calculator constants data
```

zurbergram replied:
removed

jholton commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def record_ttl
+    REDIS_CONFIG[REDIS_CONFIG_KEY.to_s]['each_ttl']
+  end
```
I believe all of this is unnecessary.  It's already taken care of for you in `Common::RedisStore#save`

zurbergram replied:
removed

jholton commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,50 @@
+  include Common::CacheAside
+
+  REDIS_CONFIG_KEY = :gi_response
+  redis_config_key REDIS_CONFIG_KEY
```
Since we don't need `REDIS_CONFIG_KEY` below, can be reduced to
`redis_config_key :gi_response`

zurbergram replied:
changed

jholton commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require 'common/models/redis_store'
+require 'common/models/concerns/cache_aside'
```
I don't think these `require`s are necessary

zurbergram replied:
rspec tests fail locally with UninitializedConstant errors

zurbergram replied:
when going to http://localhost:3000/v0/gi/calculator_constants after removing requires I get 

```
| 2019-12-18 21:19:41.345846 E [12:puma threadpool 001 sentry_logging.rb:58] [[]] Rails -- uninitialized constant Common::CacheAside.
21:19:41 web.1       | 2019-12-18 21:19:41.346031 E [12:puma threadpool 001 sentry_logging.rb:58] [[]] Rails -- /srv/vets-api/src/app/models/gids_redis.rb:5:in `<class:GIDSRedis>'
21:19:41 web.1       | /srv/vets-api/src/app/models/gids_redis.rb:4:in `<top (required)>'
21:19:41 web.1       | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:14:in `block in loading'
21:19:41 web.1       | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:151:in `exclusive'
21:19:41 web.1       | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:13:in `loading'
21:19:41 web.1       | /srv/vets-api/src/app/controllers/gids_controller.rb:9:in `service'
21:19:41 web.1       | /srv/vets-api/src/app/controllers/v0/gids/calculator_constants_controller.rb:7:in `index'```

jholton replied:
Ah, yes, looks like `cache_aside`'s namespacing doesn't match it's path for autoloading.
I think you can remove `require 'common/models/redis_store'`, but will have to keep `require 'common/models/concerns/cache_aside'`.  Seems to be working for me.

zurbergram replied:
removed

zurbergram commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def respond_to_missing?(_name, _include_private)
+    true
+  end
```
changed, though used respond_to instead as https://apidock.com/ruby/Object/respond_to_missing%3F said to not call respond_to_missing directly

jholton commented on spec/models/gids_redis_spec.rb:
```diff
@@ -0,0 +1,19 @@
+
+    expect(service.get_institution_details(scrubbed_params)).to eq(gi_response.body)
+  end
+end
```
Can we get some more spec?
Show that it caches responses.
What happens when an error occurs when reaching out to GIDS?

zurbergram replied:
Added caching test

zurbergram replied:
Service displays standard error response there is a GIDS outage otherwise return vets-api error message and stack trace of internal error.

zurbergram commented on spec/rails_helper.rb:
```diff
@@ -21,6 +21,8 @@
 require 'common/exceptions'
 require './spec/support/default_configuration_helper'
+require 'faker'
+require 'faker/medical'
```
unrelated tests were failing locally until I added these

zurbergram replied:
https://dsva.slack.com/archives/CBU0KDSB1/p1576005742268000

zurbergram commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,40 @@
+  redis_config_key :gi_response
+
+  # rest_call [Symbol] the GI::Client method to call
+  # scrubbed_params [Hash] the params to be used with the rest_call
```
changed

zurbergram commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,40 @@
+class GIDSRedis < Common::RedisStore
+  include Common::CacheAside
+
+  redis_config_key :gi_response
```
changed

zurbergram commented on lib/gi/responses/gids_response.rb:
```diff
@@ -0,0 +1,30 @@
+
+require 'common/models/redis_store'
+require 'common/client/concerns/service_status'
+require 'mvi/models/mvi_profile'
```
removed

zurbergram commented on lib/gi/responses/gids_response.rb:
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
```
moved

zurbergram commented on lib/gi/responses/gids_response.rb:
```diff
@@ -0,0 +1,30 @@
+  module Responses
+    class GIDSResponse
+      include Virtus.model(nullify_blank: true)
+      include Common::Client::ServiceStatus
```
removed

zurbergram commented on lib/gi/responses/gids_response.rb:
```diff
@@ -0,0 +1,30 @@
+      include Virtus.model(nullify_blank: true)
+      include Common::Client::ServiceStatus
+
+      attribute :status, Integer
```
added

zurbergram commented on lib/gi/responses/gids_response.rb:
```diff
@@ -0,0 +1,30 @@
+
+      attribute :status, Integer
+
+      attribute :body, Hash
```
added

zurbergram commented on lib/gi/client.rb:
```diff
@@ -2,45 +2,52 @@
 
     def get_institution_autocomplete_suggestions(params = {})
-      perform(:get, 'institutions/autocomplete', params, nil).body
+      gi_response(perform(:get, 'institutions/autocomplete', params, nil))
```
changed

jholton commented on lib/gi/gids_response.rb:
```diff
@@ -0,0 +1,27 @@
+
+module GI
+  module Responses
+    class GIDSResponse
```
Needs to be re-namespaced to:
```ruby
module GI
  class GIDSResponse
```

zurbergram replied:
renamed

zurbergram commented on app/models/gids_redis.rb:
```diff
@@ -0,0 +1,42 @@
+  attr_accessor :scrubbed_params
+
+  def method_missing(name, *args)
+    if respond_to?(name, nil)
```
removed

zurbergram commented on lib/gi/client.rb:
```diff
@@ -2,45 +2,60 @@
 
 require 'common/client/base'
 require 'gi/configuration'
+require 'gi/gids_response'
```
removed

jholton commented on lib/gi/client.rb:
```diff
@@ -1,46 +1,57 @@
 
     def get_institution_autocomplete_suggestions(params = {})
-      perform(:get, 'institutions/autocomplete', params, nil).body
+      response = perform(:get, 'institutions/autocomplete', params, nil)
```
Looks like `perform` defaults the `headers` argument to `nil`, so you don't need to specify it in your call.
```suggestion
      response = perform(:get, 'institutions/autocomplete', params)
```
Applies to all the `perform` calls in this class.

zurbergram replied:
removed

jholton commented on lib/gi/gids_response.rb:
```diff
@@ -0,0 +1,25 @@
+    # Builds a response with a ok status and a response's body
+    #
+    # @param response returned from the rest call
+    # @return [GI::Responses::GIDSResponse]
```
```suggestion
    # @return [GI::GIDSResponse]
```

jholton commented on spec/models/gids_redis_spec.rb:
```diff
@@ -0,0 +1,56 @@
+        subject.not_a_real_method
+      rescue NoMethodError => e
+        expect(e).to be_instance_of(NoMethodError)
+      end
```
Try using curly braces instead of parenthesis:
```ruby
expect{ subject.not_a_real_method }.to raise_error(NoMethodError)
```

zurbergram replied:
changed, thanks

jholton commented on spec/models/gids_redis_spec.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  describe 'method_missing' do
+    it 'returns response body' do
```
Suggested change for clarity:
```ruby
context 'when `GI::Client` responds to method' do
  it 'delegates to `GI::Client`' do
```

zurbergram replied:
changed

jholton commented on spec/models/gids_redis_spec.rb:
```diff
@@ -0,0 +1,50 @@
+      expect(subject.get_institution_details(scrubbed_params)).to eq(gids_response.body)
+    end
+
+    it 'undefined method' do
```
Suggested change for clarity:
```ruby
context 'when `GI::Client` does not respond to method' do
  it 'calls `super`' do
```

zurbergram replied:
changed

jholton commented on spec/models/gids_redis_spec.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  describe 'cached attributes' do
+    context 'get_calculator_constants' do
```
I don't think this line is necessary.  We are using `get_calculator_constants` but these specs are demonstrating caching of any method.

zurbergram replied:
removed
